### PR TITLE
Toponaming ShapeMapper

### PR DESCRIPTION
**Pull request by bgbsww**

### CODE REVIEW

**Comment by chennes** [comment]
So with that use of `std::hash` you are trying to do an injected specialization, but I don't see the implementation for `TopoDS_Shape` anywhere. You seem to be missing the 
```
template<>
struct std::hash<TopoDS_Shape>
{
    std::size_t operator()(const TopoDS_Shape& s) const noexcept
    {
        // Something like...
        return TopTools_ShapeMapHasher().HashCode(s, std::numeric_limits<StandardInteger>::max());
    }
};
```

**Comment by bgbsww** [comment]
Yeah, that didn't explain why the other patch didn't break, and why OpenCascade was recommending using it without adding that code.  But you comment got me to think and search, and the answer is that the code does exist in TopoDS_Shape.hxx, but only at 7.8.0 and above.   So I need to go back and #if version the code, because one version is required before 7.8.0, and another is required at and after it.  Thanks!

File: src/Mod/Part/App/TopoShapeMapper.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#include "TopoShapeMapper.h"
+
+namespace Part
+{
+
+void ShapeMapper::expand(const TopoDS_Shape &d, std::vector<TopoDS_Shape> &shapes)
+{
+    if (d.IsNull()) return;
+    for(TopExp_Explorer xp(d, TopAbs_FACE);xp.More();xp.Next())
+        shapes.push_back(xp.Current());
```
Comment (chennes) [inline src/Mod/Part/App/TopoShapeMapper.cpp:10]: Our clang-format file specifies that there should be curly braces around all loops and conditionals, even if they are only a single line. It's also got things to say about spacing around semicolons in for loops, etc. -- are you using an editor that can be configured to auto-apply it?

File: src/Mod/Part/App/TopoShapeMapper.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    for(TopExp_Explorer xp(d, TopAbs_EDGE, TopAbs_FACE);xp.More();xp.Next())
+        shapes.push_back(xp.Current());
+    for(TopExp_Explorer xp(d, TopAbs_VERTEX, TopAbs_EDGE);xp.More();xp.Next())
+        shapes.push_back(xp.Current());
+}
+
+void ShapeMapper::populate(bool generated,
```
Comment (chennes) [inline src/Mod/Part/App/TopoShapeMapper.cpp:17]: IMO all of these `generated` booleans should be replaced by something along the lines of
```
enum class MappingStatusOrSomething {
    GENERATED,
    MODIFIED
};
```

File: src/Mod/Part/App/TopoShapeMapper.h [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    /** Insert a map entry from a sub shape in the source to a sub shape in the new shape
+     *
+     * @params generated: whether the sub shapes are generated or modified
+     * @param s: a sub shape in the source
+     * @param d: a list of sub shapes in the new shape
+     */
+    void insert(bool generated, const TopoDS_Shape &s, const TopoDS_Shape &d);
+
+    virtual const std::vector<TopoDS_Shape> &generated(const TopoDS_Shape &s) const override {
```
Comment (chennes) [inline src/Mod/Part/App/TopoShapeMapper.h:161]:
```suggestion
    const std::vector<TopoDS_Shape> &generated(const TopoDS_Shape &s) const override {
```

File: src/Mod/Part/App/TopoShapeMapper.h [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        auto iter = _generated.find(s);
+        if(iter != _generated.end())
+            return iter->second.shapes;
+        return _res;
+    }
+
+    virtual const std::vector<TopoDS_Shape> &modified(const TopoDS_Shape &s) const override {
```
Comment (chennes) [inline src/Mod/Part/App/TopoShapeMapper.h:168]:
```suggestion
    const std::vector<TopoDS_Shape> &modified(const TopoDS_Shape &s) const override {
```

File: tests/src/Mod/Part/App/TopoShapeMapper.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    void TearDown() override
+    {
+        App::GetApplication().closeDocument(_docName.c_str());
+    }
+
+private:
+    std::string _docName;
+};
+
+TEST_F(TopoShapeMapperTest, shapeHasherTests)
```
Comment (chennes) [inline tests/src/Mod/Part/App/TopoShapeMapper.cpp:41]: I personally prefer individual test methods for each major case or set of cases that are being checked. @JohnAD what are your thoughts on that?

File: src/Mod/Part/App/TopoShapeMapper.cpp (hunk 1 above)
Comment (bgbsww) [reply src/Mod/Part/App/TopoShapeMapper.cpp:10]: Yep, VSCode will do that.   I'll set it up.  I had been shying away from modifying copied in code as much as possible; I'll bring it up to format instead.

File: src/Mod/Part/App/TopoShapeMapper.cpp (hunk 2 above)
Comment (chennes) [reply src/Mod/Part/App/TopoShapeMapper.cpp:17]: Oh, I guess in modern-ese those shouldn't be all-caps, huh? Old dog, new tricks, etc. `Generated` and `Modified` then, or some such.

File: tests/src/Mod/Part/App/TopoShapeMapper.cpp (hunk 5 above)
Comment (bgbsww) [reply tests/src/Mod/Part/App/TopoShapeMapper.cpp:41]: Yeah, I wavered on this one.   As I was writing it, it was useful to not repeat the setup and keep adding test cases, but it makes sense to split it out some now.

It seems to me that 14 individual test methods would be too many in this case; I'll aim for about 3-4 to keep the cognitive load reasonable.  That make sense, @chennes ?  Maybe a guideline number of assertions?

Comment (chennes) [reply tests/src/Mod/Part/App/TopoShapeMapper.cpp:41]: @JohnAD is the testing expert here -- I will go with the flow.

File: src/Mod/Part/App/TopoShapeMapper.h [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    };
+    typedef std::unordered_map<TopoDS_Shape, ShapeValue, ShapeHasher, ShapeHasher> ShapeMap;
+    ShapeMap _generated;
+    std::unordered_set<TopoDS_Shape, ShapeHasher, ShapeHasher> _generatedShapes;
+    ShapeMap _modified;
+    std::unordered_set<TopoDS_Shape, ShapeHasher, ShapeHasher> _modifiedShapes;
+};
+
+/// Parameters for TopoShape::makeElementFilledFace()
+struct PartExport TopoShape::BRepFillingParams
```
Comment (chennes) [inline src/Mod/Part/App/TopoShapeMapper.h:216]: Are the variable names here matching something in OCCT? It's a little Pythonic.

**Pull request merged**